vmware_guest: networks definition as a list

dagwieers opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
vmware_guest

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.3

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->
Currently the networks definition is a dict, with the network range as
key. This is problematic if the network information is coming from other
sources.

This patch turns the networks definition into a list.

This fixes #19222.

dagwieers committed:
vmware_guest: networks definition as a list

Currently the networks definition is a dict, with the network range as
key. This is problematic if the network information is coming from other
sources.

This patch turns the networks definition into a list.

This fixes #19222.

dagwieers commented:
@nerzhul @angystardust @dav1x Please test !

dav1x commented:
@dagwieers ill test today! I've got a PR i need to submit on this for multiple static IP addresses set as well!

dagwieers commented:
ready_for_review

nerzhul commented:
The concept is interesting and i want to change dict to list too, but i haven't tested it yet, will take some time this afternoon

angystardust commented:
@dagwieers It's ok for me but after the merge you ask @aperigault to rebase https://github.com/ansible/ansible/pull/19866

dav1x commented:
@dagwieers 

I'm having problems getting this to work unless I change the module param type to a list for networks:

Here is my playbook:
```
---
- name: Clone template and customize
  hosts: localhost
  tasks:
    - vmware_guest:
        hostname: 10.x.x.25
        username: administrator@vsphere.local
        password: password
        validate_certs: False
        name: testvm-2
        datacenter: boston
        cluster: devel
        validate_certs: False
        template: ansible-template
        customize: True                                                                                                                                                                                 
        domain: example.com
        dns_servers: ['192.168.1.1','192.168.1.2']
        networks:
        - name: 'VM Network'
          network: '192.168.1.0/24'
          ip: '192.168.1.100'
          gateway: '192.168.1.1'

```

I also noticed that the getFolder function is case sensitive and may be confusing for customers. Ill open an issue for that.  Beyond having to modify works great!

dav1x commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -1649,7 +1651,7 @@ def main():
             dns_servers=dict(required=False, type='list', default=None),
             domain=dict(required=False, type='str', default=None),
-            networks=dict(required=False, type='dict', default={}),
+            networks=dict(required=False, type='dict', default=[]),
```
networks=dict(required=False, type='list', default=[]),

dagwieers replied:
Ah, good catch. You won't believe it, but the whole patch was typed from one isolated system to my laptop :-) Don't ask...

dagwieers replied:
And fixed ! Thanks !

nerzhul commented:
i will test it tomorrow or on monday on my infrastructure to test idempotency on VM from scratch

dagwieers commented:
ready_for_review (again)

aperigault commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -883,34 +885,36 @@ def configure_network(self, vm_obj):
-                    else:
-                        self.module.fail_json(msg="VLAN %(vlan)s doesn't exists" % self.params['networks'][network])
+        for network in self.params.get('networks', list()):
+            if 'network' not in network:
```
Network is not mandatory. Is mandatory only if you set ip param.
Furthermore for me network field is useless. This is not the vmware_guest job to check if you set wrong ip (not in good subnet) on your interface but it's not realy a problem.

dagwieers replied:
The old implementation gets the netmask from the network information. And that's how I was using it as well. Sometimes it is easier to provide the CIDR information instead of calculating the netmask.
(I don't mind, I know colleagues do though)

Let's see what we want to do with this. I am open to ideas.

aperigault replied:
My idea:
if network is set use it like it is already implemented (subnet check + netmask calcul)
If netmask is set use it instead of calculating it
I also imagine to set ip + cidr in ip field and use it to calcul netmask (if netmask is not set).
With this everyone will be happy? No?

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -883,34 +885,36 @@ def configure_network(self, vm_obj):
-                    else:
-                        self.module.fail_json(msg="VLAN %(vlan)s doesn't exists" % self.params['networks'][network])
+        for network in self.params.get('networks', list()):
+            if 'network' not in network:
```
You're right. I don't know where that comes from TBH, I think it was there before though.
I will refactor that part.

aperigault commented:
My idea:
if network is set use it like it is already implemented (subnet check + netmask calcul)
If netmask is set use it instead of calculating it
I also imagine to set ip + cidr in ip field and use it to calcul netmask (if netmask is not set).
With this everyone will be happy? No?

aperigault commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -883,34 +888,44 @@ def configure_network(self, vm_obj):
+                if 'netmask' in network:
+                    pass
+
+                # TODO: What if netmask and network are provided but conflict ?
```
If you want you can add an error: "You can't define netmask and network at the same time" or simply ignore network when netmask is define like it's already implemented ?

dagwieers replied:
Well, I would rather make the interface more simple and not give the choice.
I think @nerzhul suggested this earlier as well, we both prefer a CIDR over a netmask.

aperigault commented:
LGTM

nerzhul commented:
maybe force CIDR usage ?

dagwieers commented:
That's what I prefer.

dav1x commented:
The customize param is deprecated correct?

https://github.com/dagwieers/ansible/blob/f7fef00848d980400570a4060646b64e1856bca6/lib/ansible/modules/cloud/vmware/vmware_guest.py#L1333-L1346

and 

https://github.com/dagwieers/ansible/blob/f7fef00848d980400570a4060646b64e1856bca6/lib/ansible/modules/cloud/vmware/vmware_guest.py#L1785

I'm having issues getting this to work. Additionally, the networks param option is converted back to a dict.

dagwieers commented:
Yeah, I need to rebase again. Normally ansibot is telling me though ;-)

dagwieers committed:
Remove the option to provide either netmask or network

It is more concise if there's only one way to configure it.
So provide both `ip` and `network`.

(I also re-indented a section due to a useless check)

dav1x commented:
I've only tested the linux portion of the customization. IP customization and domain, dns_servers, dns_suffix are all working great with this slight modification:

```
1001c1001
<                 if 'ip' in self.params['networks'][network]:
---
>                 if 'ip' in network:
```

aperigault commented:
If we want to simplify why a network param ? Ip: ip/cidr is not more simple ?

dagwieers commented:
@aperigault Yes, that's even better if people understand that format correctly. I am willing to change it to that proposal if everyone agrees. @nerzhul @dav1x @angystardust 

Do we think this:
```yaml
  networks:
  - name: VM Network
    ip: 192.168.1.13/24
```
is better than this:
```yaml
  networks:
  - name: VM Network
  - ip: 192.168.1.13
    network: 192.168.1.0/24
```

dagwieers commented:
@dav1x Argh, another artifact of one of the many rebases :-( It's been a bit too much.

dav1x commented:
@dagwieers @aperigault 

I like the simplification of dropping network. But, I agree with @dagwieers concerning the user expecting ip: to be only an ip address.

I would imagine in playbooks and such inventory's and calls would be much easier to dynamically generate without the overhead of appending a netmask CIDR suffix to the address. 

Thoughts @jctanner @nerzhul ?

dagwieers commented:
@dav1x We do have a few jinja filters for network handling/mangling, however the stuff we need is not exposed externally :-( If they would be available, the question is whether we expect our users to have to go to that level of understanding. We could add examples though, not sure if that is going to be sufficient.

I am open to that and extending the exposed filters to do the required mangling from IP/network or IP/netmask -> CIDR. If we consider this a good idea.

nerzhul commented:
maybe a couple ip/netmask attribute could be sufficient and clear for every user

aperigault commented:
Maybe i'm bad but if you don't know what is a netmask or a cidr just use dhcp...

dagwieers commented:
I am open to whatever is decided.

*Option 1*
```yaml
  networks:
    ip: 192.168.1.2
    netmask: 255.255.255.0
```
*Option 2*
```yaml
  networks:
    ip: 192.168.1.2/24
```
*Option 3*
```yaml
  networks:
    ip: 192.168.1.2
    network: 192.168.1.0/24
```
Or simply accept every combination, and escalate conflicted information.

- Option 1: Is what most people will expect or have available
- Option 2: Is concise and minimal 
- Option 3: Sometimes VLAN information is provided in CIDR notation

My preference is **Option 2** + proper jinja filters + good examples, but that's from a technical perspective and a let-us-show-them-the-proper-way-to-do-it. But then again, from a practical point, **Option 1** and **Option 3** are probably preferred.

Let the games begin ! @bcoca Your view on this is very much appreciated as well.

dav1x commented:
The Linux admin in me says option 1. But, I agree with nerzhul, im okay with 1 or 2.

dagwieers commented:
Ok, let me make sure we also have all the imaginable network-mangling jinja2 filters anyway (so Option 1 and Option 3 becomes possible indirectly) and I will implement Option 2. Thanks everyone for playing !

I will already modify this PR so it becomes mergable ASAP, and work on the network-mangling jinja2 filters.

dagwieers commented:
ready_for_review

dav1x commented:
line 889 has a logic error missing not for netmask not in network.
    if 'ip' in network and 'netmask' in network:                                                                                                                                       
                    self.module.fail_json(msg="Both 'ip' and 'network' are required together.")

Also in your documentation you changed the network var name to name and thoughout the module its referred to as both.

dagwieers committed:
Fix bugs reported by @dav1x

Thanks !

dagwieers commented:
@dav1x @nerzhul Please re-review !

dav1x commented:
tested and works great! shipit
